### PR TITLE
Immutable samplers

### LINKED ISSUE #1723: Immutable samplers
Vulkan allows defining immutable samplers on descriptor set layout creation.
Need to investigate how to map these to other backends.

### PR DESCRIPTION
**Pull request by kvark**
Fixes #1723
- [x] hal/empty/render/warden
- [x] metal
- [ ] d3d12
- [x] vulkan

PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends:

### CODE REVIEW

**Comment by kvark** [comment]
I don't want to ship the DX11/12 support with this PR, can go as follow uo

File: src/hal/src/pso/descriptor.rs [hunk 1]
```diff
@@ -74,7 +74,8 @@ pub struct DescriptorSetLayoutBinding {
     pub count: DescriptorArrayIndex,
     /// Valid shader stages.
     pub stage_flags: ShaderStageFlags,
-    // TODO: immutable samplers?
+    /// Use the associated list of immutable samplers.
+    pub immutable_samplers: bool,
```
Comment (msiglreith) [inline src/hal/src/pso/descriptor.rs]: where is this needed?

Comment (msiglreith) [reply src/hal/src/pso/descriptor.rs]: Resolved in internal discussion.
Vulkan takes a null ptr or array of samplers. Works basically like a Option, but combining the samplers into one iterator.

File: src/backend/vulkan/src/device.rs [hunk 2]
```diff
@@ -1113,7 +1121,13 @@ impl d::Device<B> for Device {
                 descriptor_type: conv::map_descriptor_type(b.ty),
                 descriptor_count: b.count as _,
                 stage_flags: conv::map_stage_flags(b.stage_flags),
-                p_immutable_samplers: ptr::null(), // TODO
+                p_immutable_samplers: if b.immutable_samplers {
+                    let slice = &immutable_samplers[sampler_offset..];
```
Comment (grovesNL) [inline src/backend/vulkan/src/device.rs]: Why don't we need `sampler_offset + b.count` here to match the others?

**Comment by kvark** [comment]
Because all we care about here is a pointer

**Comment by kvark** [comment]
Thanks for the reviews!
Appears to be no blockers.
Bors r=msiglreith,grovesNL

**Pull request merged**